WcaOnRails/app/controllers/database_controller.rb

```diff
@@ -4,38 +4,41 @@ class DatabaseController < ApplicationController
   RESULTS_README_TEMPLATE = 'database/public_results_readme'
 
   def results_export
-    @sql_filename, @sql_filesize = _link_info DbDumpHelper::RESULTS_EXPORT_SQL_PERMALINK
-    @tsv_filename, @tsv_filesize = _link_info DbDumpHelper::RESULTS_EXPORT_TSV_PERMALINK
+    @sql_path, @sql_filesize = get_current_export("sql")
+    @tsv_path, @tsv_filesize = get_current_export("tsv")
+    @sql_filename = File.basename(@sql_path)
+    @tsv_filename = File.basename(@tsv_path)
 
-    @sql_rel_path = DatabaseController.rel_download_path DbDumpHelper::RESULTS_EXPORT_FOLDER, @sql_filename
-    @tsv_rel_path = DatabaseController.rel_download_path DbDumpHelper::RESULTS_EXPORT_FOLDER, @tsv_filename
-
-    @sql_perma_path = DatabaseController.rel_download_path DbDumpHelper::RESULTS_EXPORT_FOLDER, DbDumpHelper::RESULTS_EXPORT_SQL_PERMALINK
-    @tsv_perma_path = DatabaseController.rel_download_path DbDumpHelper::RESULTS_EXPORT_FOLDER, DbDumpHelper::RESULTS_EXPORT_TSV_PERMALINK
+    @sql_perma_path = "#{EnvConfig.ROOT_URL}/export/results/#{DbDumpHelper::RESULTS_EXPORT_SQL_PERMALINK}"
+    @tsv_perma_path = "#{EnvConfig.ROOT_URL}/export/results/#{DbDumpHelper::RESULTS_EXPORT_TSV_PERMALINK}"
   end
 
-  def _link_info(permalink)
-    full_permalink = DbDumpHelper::RESULTS_EXPORT_FOLDER.join(permalink)
-
-    actual_filename = File.basename File.readlink(full_permalink)
-    # Manually resolve the relative link to be independent of runtime environments
-    actual_file = DbDumpHelper::RESULTS_EXPORT_FOLDER.join(actual_filename)
-
-    filesize_bytes = File.size? actual_file
-
-    [actual_filename, filesize_bytes]
+  def sql_permalink
+    url, _ = get_current_export("sql")
+    redirect_to url, status: 301, allow_other_host: true
   end
 
-  def developer_export
-    @rel_download_path = DatabaseController.rel_download_path DbDumpHelper::DEVELOPER_EXPORT_FOLDER, DbDumpHelper::DEVELOPER_EXPORT_SQL_PERMALINK
+  def tsv_permalink
+    url, _ = get_current_export("sql")
+    redirect_to url, status: 301, allow_other_host: true
   end
 
-  def self.rel_download_path(base_folder, file_name)
-    file_path = base_folder.join file_name
-    relative_path = file_path.relative_path_from DbDumpHelper::EXPORT_PUBLIC_FOLDER.parent
+  def get_current_export(type)
+    export_timestamp = Time.new(DbDumpHelper::export_metadata["export_date"])
+
+    Rails.cache.fetch("database-export-#{export_timestamp}-#{type}", expires_in: 2.days) do
+      file_name = "#{DbDumpHelper::RESULTS_EXPORT_FOLDER}/WCA_export#{export_timestamp.strftime('%j')}_#{export_timestamp.strftime('%Y%m%dT%H%M%SZ')}.#{type}.zip"
+      bucket = Aws::S3::Resource.new(
+        region: EnvConfig.STORAGE_AWS_REGION,
+        credentials: Aws::InstanceProfileCredentials.new,
+        ).bucket(DbDumpHelper::BUCKET_NAME)
+      filesize_bytes = bucket.object(file_name).content_length
+      ["https://s3.#{EnvConfig.AWS_STORAGE_REGION}.amazonaws.com/#{DbDumpHelper::BUCKET_NAME}/#{file_name}", filesize_bytes]
```

Comment:
Out of curiosity: Am I reading it correctly that amazon provides human-guessable permalinks to every public S3 bucket in the whole world? And does this imply that our `DbDumpHelper::BUCKET_NAME` has to be unique among the whole us-west-2 region?!

Reply:
yes, it even has to be globally unique (crazy right!)